tools/update-tester/src/parser.rs

```diff
@@ -0,0 +1,447 @@
+use std::{collections::HashMap, ffi::OsStr, fs};
+
+use pulldown_cmark::{
+    CodeBlockKind::Fenced,
+    CowStr, Event, Parser,
+    Tag::{CodeBlock, Heading},
+};
+use semver::Version;
+
+#[derive(Debug, PartialEq, Eq)]
+#[must_use]
+pub struct TestFile {
+    name: String,
+    stateless: bool,
+    pub tests: Vec<Test>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[must_use]
+pub struct Test {
+    pub location: String,
+    pub header: String,
+    pub text: String,
+    pub output: Vec<Vec<String>>,
+    transactional: bool,
+    ignore_output: bool,
+    pub precision_limits: HashMap<usize, usize>,
+    pub creation: bool,
+    pub validation: bool,
+    pub min_toolkit_version: Option<Version>,
+}
+
+pub fn extract_tests(root: &str) -> Vec<TestFile> {
+    // TODO handle when root is a file
+    let mut all_tests = vec![];
+    let walker = walkdir::WalkDir::new(root)
+        .follow_links(true)
+        .sort_by(|a, b| a.path().cmp(b.path()));
+    for entry in walker {
+        let entry = entry.unwrap();
+        if !entry.file_type().is_file() {
+            continue;
+        }
+
+        if entry.path().extension() != Some(OsStr::new("md")) {
+            continue;
+        }
+
+        let realpath;
+        let path = if entry.file_type().is_symlink() {
```

Comment:
This is superfluous:  just pass `entry.path()` directly to `read_to_string`.

Reply:
👍 Would you like me to also update this line in sql-doctester?